hardhat/README.md

```diff
@@ -0,0 +1,236 @@
+
+1. Install the Optimistic Ethereum package in the application.
+   ```sh
+   yarn add @eth-optimism/hardhat-ovm
+   ```
+2. Edit `hardhat.config.js` to use the Optimistic Ethereum package.
+   ```js
+   require("@nomiclabs/hardhat-waffle");
+   require('@eth-optimism/hardhat-ovm')
+
+   ...
+   ```
+3. In the same file, add `optimistic` to the list of networks:
+   ```js
+   // hardhat.config.js
+
+   ...
+   
+   module.exports = {
+     solidity: "0.8.4",   // or whatever you get from HardHat
+     networks: {
+       // Add this network to your config!
+       optimistic: {
+          url: 'http://127.0.0.1:8545',
+          accounts: {
+             mnemonic: 'test test test test test test test test test test test junk'
+          },
+          // This sets the gas price to 0 for all transactions on L2. We do this
+          // because account balances are not automatically initiated with an ETH
+          // balance (yet, sorry!).
+          gasPrice: 0,
+          ovm: true // This sets the network as using the ovm and ensure contract will be compiled against that.
+       }
+     }
+   }
+   ```
+
+   At this point you need to wait until the `docker-compose build` ends, if it hasn't yet, and then run
+   `cd ~/optimism/ops ; docker-compose up`.
+
+4. Test the contract on Optimistic Ethereum. Hardhat will recognize it has not been compiled and compile it for you.
+
+   ```sh
+   npx hardhat --network optimistic test
+   ```
+
+5. If you want to interact with the app manually, use the console. You can use the same JavaScript commands
+   to control it you used above.
+   ```sh
+   npx hardhat --network optimistic console
+   ```
+   
+   
+#### Change the Solidity Version if Needed
+
+To run on Optimistic Ethereum a contract needs to be compiled with a variant Solidity compiler. Sometimes
+the latest version of Solidity supported by Optimistic Ethereum is not the same as the version used by the
+sample app in HardHat. When that is the case, the `npx hardhat --network optimistic test` command
+fails with an error message similar to:
+
+```
+OVM Compiler Error (insert "// @unsupported: ovm" if you don't want this file to be compiled for the OVM):
+ contracts/Greeter.sol:2:1: ParserError: Source file requires different compiler version (current compiler is 0.7.6) - note that nightly builds are considered to be strictly less than the released version
+pragma solidity ^0.8.0;
+^---------------------^
+
+Error HH600: Compilation failed
+```
+
+To solve this problem:
+
+1. Edit the `hardhat.config.js` file to change `module.exports.solidity` to the supported version.
+2. Edit `contracts/Greeter.sol` to change the `pragma solidity` line to the supported version.
+3. Check the application still works on normal Ethereum.
+   ```sh
+   npx hardhat test
+   ```
+4. Check the application works on Optimistic Ethereum.
+   ```sh
+   npx hardhat --network optimistic test
+   ```
+
+
+## How to Run Tests
```

Comment:
Hmmm section title is a little confusing. Maybe something like "Best practices for running tests".